File: java/src/main/java/challenges/aoc2024/day5/function/PageOrderingComparator.java
package challenges.aoc2024.day5.function;

import java.util.Comparator;
import java.util.List;
import java.util.Objects;
import java.util.Optional;

import challenges.aoc2024.day5.model.PageOrderingRule;

/**
 * Comparator class for comparing two page numbers according to the ordering rules.
 */
public class PageOrderingComparator implements Comparator<Integer> {
    private final List<PageOrderingRule> rules;

    /**
     * Comparator class for comparing two page numbers according to the ordering rules.
     * 
     * @param rules The rules to order pages by.
     */
    public PageOrderingComparator(List<PageOrderingRule> rules) {
        this.rules = rules;
    }

    /**
     * Compare two page numbers according to the ordering rules.
     */
    @Override
    public int compare(Integer firstPage, Integer secondPage) {
        final Optional<PageOrderingRule> relevantRule = rules.stream()
            .filter(rule -> isRelevantRule(rule, firstPage, secondPage))
            .findFirst();

        if (!relevantRule.isPresent()) {
            return 0;
        }

        return Objects.equals(firstPage, relevantRule.get().getBeforeNumber()) ? -1 : 1;
    }

    /**
     * Determine if a specific ordering rule is relevant to the numbers that are being compared.
     * 
     * @param rule The rule to evaluate.
     * @param firstPage The first page number.
     * @param secondPage The second page number.
     * 
     * @return True if the rule is relevant to the numbers that are being compared.
     */
    private boolean isRelevantRule(PageOrderingRule rule, Integer firstPage, Integer secondPage) {
        
        return (Objects.equals(rule.getBeforeNumber(), firstPage) && Objects.equals(rule.getAfterNumber(), secondPage))
            || (Objects.equals(rule.getBeforeNumber(), secondPage) && Objects.equals(rule.getAfterNumber(), firstPage));
    }
    
}

File: scala/src/main/scala/challenges/aoc2024/day5/function/PageOrderingComparator.scala
package challenges.aoc2024.day5.function

import challenges.aoc2024.day5.model.PageOrderingRule

class PageOrderingComparator(rules: List[PageOrderingRule]) extends Ordering[Int] {

  override def compare(firstPage: Int, secondPage: Int): Int = {
    try
      val relevantRule: PageOrderingRule = rules
        .filter(rule => isRelevantRule(rule, firstPage, secondPage))
        .head

      if firstPage == relevantRule.beforeNumber then -1 else 1

    catch
      case e: NoSuchElementException => 0
  }

  private def isRelevantRule(rule: PageOrderingRule, firstPage: Int, secondPage: Int): Boolean = {
    (rule.beforeNumber == firstPage && rule.afterNumber == secondPage)
    || (rule.beforeNumber == secondPage && rule.afterNumber == firstPage)
  }
}
